File: Lab5_Java/src/Company.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Company implements Cloneable
{
    private String brand;
    private Devices devices;
    private Creation creation;
    public Company(String brand, Devices devices, Creation creation)
    {
        this.brand = brand;
        this.devices = devices;
        this.creation = creation;
    }
    public Company()
    {
        this.devices = new Devices();
        this.creation = new Creation();
        this.brand = "Brand";
    }
    public void input()
    {
        int n_sp, n_lt;
        Scanner inp = new Scanner(System.in);
        System.out.printf("Введите название компании ");
        brand = inp.next();
        System.out.printf("Введите число моделей смартфонов вашей компании ");
        n_sp = inp.nextInt();
        devices.setN_sp(n_sp);
        System.out.printf("Введите число моделей ноутбуков вашей компании ");
        n_lt = inp.nextInt();
        devices.setN_lp(n_lt);
        devices.input_dev();
        creation.input_creat();
    }
    public void output()
    {
        System.out.printf("Название компании: %s\n", brand);
        this.devices.output_dev();
        this.creation.output_creat();
    }
    public void outputstr()
    {
        devices.outstr_dev();
    }
    public float income()
    {
        float income, taxes;
        taxes = creation.get_taxes();
        income = (float)devices.income_dev() * (100 - taxes) / 100;
        return income;
    }
    public int expenses()
    {
        return creation.expenses_creat();
    }
    public float profit()
    {
        return income() - expenses();
    }

    public void writefile()
    {
        Scanner inp = new Scanner(System.in);
        System.out.printf("Введите название файла ");
        String namefile = inp.next();
        File file = new File(namefile);
        try {
            //проверяем, что если файл не существует то создаем его
            if(!file.exists()){
                file.createNewFile();
            }

            //PrintWriter обеспечит возможности записи в файл
            PrintWriter out = new PrintWriter(file.getAbsoluteFile());

            try
            {
                out.print(brand + "\n");
                out.close();
                devices.writefile(namefile);
            } finally {
                out.close();
            }
        } catch(IOException e) {
            throw new RuntimeException(e);
        }
    }
    public Object clone()  // перегрузка интерфейса клонирования
    {
        try
        {
            Company clone=(Company)super.clone();
            clone.devices=(Devices)devices.clone();
            clone.creation=(Creation)creation.clone();
            return clone;
        }
        catch(CloneNotSupportedException e)
        {
        }
        return this;
    }
}

File: Lab5_C++/company.cpp
#define _CRT_SECURE_NO_WARNINGS
#include "company.h"
#include <iostream>
#include <string>
#include <fstream>
using namespace std;

Company::Company(const Company& company)
{
    brand = string(company.brand);
    devices = Devices(company.devices);
    creation = Creation(company.creation);
}

Company::Company(string brand, Devices devices, Creation creation)
{
    this->brand = brand;
    this->devices = devices;
    this->creation = creation;
}

Company::Company()
{
    devices = Devices();
    creation = Creation();
    brand = "Brand";
}

Company::Company(string brand)
{
    this->brand = brand;
    devices = Devices();
    creation = Creation();
}

Company::~Company() {}

void Company::input(int* n_sp, int* n_lt)
{
    cout << "??????? ???????? ????????" << endl;
    cin >> brand;
    cout << "??????? ????? ??????? ?????????? ????? ????????" << endl;
    cin >> *n_sp;
    cout << "??????? ????? ??????? ????????? ????? ????????" << endl;
    cin >> *n_lt;
    devices.input_dev(*n_sp, *n_lt);
    creation.input_creat();
}

void Company::output(int n_sp, int n_lt)
{
    cout << brand << endl;
    devices.output_dev();
    creation.output_creat();
}

void Company::outputstr(int n_sp, int n_lt)
{
    devices.outstr_dev();
}

float Company::income(int n_sp, int n_lt)
{
    float income;
    income = devices.income_dev() * (100 - creation.taxes) / 100;
    return income;
}

int Company::expenses()
{
    return creation.expenses_creat();
}

int Company::profit(int n_sp, int n_lt)
{
    return income(n_sp, n_lt) - expenses();
}

void Company::writefile()
{
    string namefile;
    cout << "??????? ???????? ?????" << endl;
    cin >> namefile;
    ofstream fout(namefile);
    try
    {
        if (!fout.is_open())
            throw 1;
        fout.open(namefile, ios_base::trunc);
        fout << brand << endl;
        fout.close();
        devices.writefile(namefile);
    }
    catch (...)
    {
        cout << "?????? ???????? ?????" << endl;
    }
}